### PR TITLE
Make mode token aggregation deterministic

### PR DESCRIPTION
**Pull request by guillaume-be**
- Addresses non-deterministic behaviour identified in https://github.com/guillaume-be/rust-bert/pull/252 where the label of a token made of sub-tokens with an equal amount of most-common labels (e.g. 1 `O` 1`B`) was selected randomly.
- Now assigns the label that reached the highest score in case of most common label tie

### CODE REVIEW

File: src/pipelines/token_classification.rs [hunk 1]
```diff
@@ -1129,12 +1130,18 @@ impl TokenClassificationModel {
...
+                    *entry = (
+                        entry.0 + 1,
+                        if c.score > entry.1 { c.score } else { entry.1 },
+                    );
                     m
                 });
                 counts
                     .into_iter()
-                    .max_by(|a, b| a.1.cmp(&b.1))
+                    .max_by_key(|item| (item.1 .0, OrderedFloat(item.1 .1)))
```
Comment (mlodato517) [inline src/pipelines/token_classification.rs:1144]: I found this `.1 .0` syntax pretty hard to read - I wonder if destructuring here would be easier to read (and symmetric with the `map` on the next line):
```suggestion
                    .max_by_key(|&(_, (count, score))| (count, OrderedFloat(score)))
```

File: src/pipelines/token_classification.rs [hunk 2]
```diff
@@ -1129,12 +1130,18 @@ impl TokenClassificationModel {
...
             LabelAggregationOption::Mode => {
                 let counts = tokens.iter().fold(HashMap::new(), |mut m, c| {
-                    *m.entry((c.label_index, c.label.as_str())).or_insert(0) += 1;
+                    let entry = m
+                        .entry((c.label_index, c.label.as_str()))
+                        .or_insert((0, 0.));
+                    *entry = (
+                        entry.0 + 1,
+                        if c.score > entry.1 { c.score } else { entry.1 },
+                    );
```
Comment (mlodato517) [inline src/pipelines/token_classification.rs:1139]: I think we can use [`max`]() here to clean up the second part of that tuple. And then, so we don't have to remember what `.0` and `.1` are we can maybe destructure the entry at the beginning:
```suggestion
                    let (ref mut count, ref mut score) = m
                        .entry((c.label_index, c.label.as_str()))
                        .or_insert((0, 0.0_f64));
                    *count += 1;
                    *score = score.max(c.score);
```

**Review by mlodato517 — COMMENTED** [review]
My approval here probably wouldn't mean anything but this makes sense to me! I wonder if it's worth writing a unit test for (the test would only have caught the previous behavior X% of the time so it's not obvious it's a good idea but a test that fails X% of the time would maybe have indicated an opportunity for improvement earlier on? :shrug: ). Other than that, just saw some little places to name some variables more clearly (in my opinion!).

**Comment by guillaume-be** [comment]
> My approval here probably wouldn't mean anything but this makes sense to me! I wonder if it's worth writing a unit test for (the test would only have caught the previous behavior X% of the time so it's not obvious it's a good idea but a test that fails X% of the time would maybe have indicated an opportunity for improvement earlier on? 🤷 ). Other than that, just saw some little places to name some variables more clearly (in my opinion!).

Thank you for the suggestion - indeed destructuring is cleaner!

**Pull request merged**